File: krypto/02/knapsack-crypto/src/main/java/algorithm/MHCipher.java
package algorithm;

import exception.MessageException;
import lombok.Data;

@Data
public class MHCipher {
    private SimpleKeyGenerator keyGen;
    private long[] privateKey, publicKey;

    public MHCipher(SimpleKeyGenerator keyGen) {
        this.keyGen = keyGen;
        generateKeys();
    }

    private void generateKeys() {
        privateKey = keyGen.generateDefaultPrivateKey();
        publicKey = keyGen.generatePublicKey(privateKey);
    }

    public void generatePublicKey() {
        publicKey = keyGen.generatePublicKey(privateKey);
    }

    public long[] encrypt(String message) throws MessageException {
        int length = message.length();
        if (length <= 0)
            throw new MessageException("Message is empty");

        long[] output = new long[length];
        for (int i = 0; i < length; i++) {
            output[i] = calculateEncryptedChar(message.charAt(i));
        }
        return output;
    }

    public String decrypt(long[] cipher) {
        StringBuilder stringBuilder = new StringBuilder();
        long multiInverse = calculateMultiplierModuloInverse();

        for (long element : cipher) {
            long encryptedChar = (multiInverse * element) % keyGen.getModulus();
            stringBuilder.append(decryptChar(encryptedChar));
        }

        return stringBuilder.toString();
    }

    private char decryptChar(long encryptedChar) {
        int MSB = 0x80;
        int outputChar = 0;
        for (int i = 7; i >= 0 && encryptedChar > 0; i--) {
            if (encryptedChar >= privateKey[i]) {
                encryptedChar -= privateKey[i];
                outputChar += MSB;
            }
            MSB /= 2;
        }
        return (char) outputChar;
    }

    private long calculateMultiplierModuloInverse() {
        long[] response = extendedEuclid(keyGen.getMultiplier(), keyGen.getModulus());
        return response[0];
    }

    private long[] extendedEuclid(long c, long d) {
        if (d == 0)
            return new long[]{1, 0};

        long[] response = extendedEuclid(d, c % d);
        long a = response[1];
        long b = response[0] - (c / d) * response[1];
        return new long[]{a, b};
    }

    private long calculateEncryptedChar(char x) {
        long output = 0;
        for (int i = 0; i < 8; i++) {
            if (isOdd(x)) {
                output += publicKey[i];
            }
            x /= 2;
        }
        return output;
    }

    private boolean isOdd(char x) {
        return x % 2 != 0;
    }
}

File: krypto/zad2py/mh_cipher.py
from simple_key_generator import SimpleKeyGenerator


class MHCipher:
    def __init__(self):
        self.keygen = SimpleKeyGenerator()
        self.private_key = None
        self.public_key = None
        self.generate_keys()

    def generate_keys(self):
        self.private_key = self.keygen.generate_default_private_key()
        self.public_key = self.keygen.generate_default_public_key(self.private_key)

    def generate_public_key(self):
        self.public_key = self.keygen.generate_default_public_key(self.public_key)

    def encrypt(self, message):
        message_length = len(message)

        if message_length == 0:
            return

        output = []
        for i in message:
            output.append(self.calculate_encrypted_char(i))
        return output

    def decrypt(self, cipher):
        output = ""
        multi_inverse = self.calculate_multiplier_module_inverse()
        print(multi_inverse)

        for i in cipher:
            encrypted_char = (multi_inverse * i) % self.keygen.modulus
            output += self.decrypt_char(encrypted_char)
        return output

    def decrypt_char(self, encrypted_char):
        msb = 128
        output_char = 0
        for i in range(7, -1, -1):
            if encrypted_char <= 0:
                break
            if encrypted_char >= self.private_key[i]:
                encrypted_char -= self.private_key[i]
                output_char += msb
            msb /= 2
        return chr(int(output_char))

    def calculate_encrypted_char(self, char):
        output = 0
        ascii_value = ord(char)
        for i in range(8):
            if ascii_value % 2 != 0:
                output += self.public_key[i]
            ascii_value /= 2
            ascii_value = int(ascii_value)
        return output

    def extended_euclid(self, c, d):
        if d == 0:
            return [1, 0]

        response = self.extended_euclid(d, c % d)
        a = response[1]
        b = response[0] - (int(c / d) * response[1])
        return [a, b]

    def calculate_multiplier_module_inverse(self):
        # response = self.extended_euclid(self.keygen.multiplier, self.keygen.modulus)
        # return response[0]
        return self.egcd(self.keygen.multiplier, self.keygen.modulus)

    def egcd(self, a, b):
        x, y, u, v = 0, 1, 1, 0
        while a != 0:
            q, r = b // a, b % a
            m, n = x - u * q, y - v * q
            b, a, x, y, u, v = a, r, u, v, m, n
        gcd = b
        return x
